compiler/hash-typecheck/src/ops/validate.rs

```diff
@@ -1192,4 +1194,56 @@ impl<'tc> Validator<'tc> {
             _ => Ok(None),
         }
     }
+
+    /// Validate that a [Pat::Range] has matching `lo`, and `hi` types, and that
+    /// the values of ranges are semantically correct
+    pub(crate) fn validate_range_pat(&mut self, pat: &RangePat) -> TcResult<()> {
+        let RangePat { lo, hi, end } = *pat;
+
+        // Expect to be literal terms, which should definitely unify
+        let hi_ty = self.typer().infer_ty_of_term(hi)?;
+        let lo_ty = self.typer().infer_ty_of_term(lo)?;
+        let _ = self.unifier().unify_terms(hi_ty, lo_ty)?;
```

Comment:
This should be a bi-directional unification

Reply:
Should it use `terms_are_equal()`

Reply:
Yeah, maube create `unify_terms_bi` which is the same as terms_are_equal but retains the TcResult.

Reply:
Should it discard the substitutions or apply them on the terms?

Reply:
the unify_terms_bi should merge the two subs and return them. the validator should ensure empty.